packages/entity-server/src/routes/hierarchy/EntitySearchRoute.ts

```diff
@@ -0,0 +1,63 @@
+/**
+ * Tupaia
+ * Copyright (c) 2017 - 2023 Beyond Essential Systems Pty Ltd
+ */
+
+import { Route } from '@tupaia/server-boilerplate';
+import { QUERY_CONJUNCTIONS } from '@tupaia/database';
+import { formatEntitiesForResponse } from './format';
+import {
+  MultiEntityRequest,
+  MultiEntityRequestParams,
+  MultiEntityRequestBody,
+  EntityRequestQuery,
+  EntityResponse,
+} from './types';
+
+const { RAW } = QUERY_CONJUNCTIONS;
+
+export type EntitySearchRequest = MultiEntityRequest<
+  MultiEntityRequestParams & { searchString: string },
+  EntityResponse[],
+  MultiEntityRequestBody,
+  EntityRequestQuery & { pageSize?: number; page?: number }
+>;
+export class EntitySearchRoute extends Route<EntitySearchRequest> {
+  public async buildResponse() {
+    const { hierarchyId, fields, field, filter } = this.req.ctx;
+    const { searchString: rawString } = this.req.params;
+    const { pageSize, page } = this.req.query;
+    const searchString = rawString.toLowerCase();
+
+    const entities = await this.req.models.entity.find(
+      {
+        ...filter,
+        [RAW]: {
+          // Running this RAW is much faster for larger hierarchies
```

Comment:
Anecdotally, about 6x quicker than an array filter

Reply:
Oooh interesting. I wonder if this is more performant than the existing hierarchy functions on `EntityModel` (`getDescendantsOfEntities`,`getAncestorsOfEntities`,etc.) I think those use a join on the `ancestor_descendant_relation` table

Reply:
It'd require some testing but academically the answer is "probably not". In general postgres can optimise a join query better than a subquery.